returnn/tf/util/data.py

```diff
@@ -914,6 +914,11 @@ def declare_same_as(self, other):
       self_same_as.declare_same_as(other_same_base)
       if (self.dyn_size_ext is None or not self._validate_in_current_graph()) and self_same_as.dyn_size_ext:
         self.dyn_size_ext = self_same_as.get_dyn_size_ext_for_batch_ctx(self.batch, self.control_flow_ctx)
+    self_derived_bases = set(self.get_derived_bases_list())
+    other_derived_bases = set(other.get_derived_bases_list())
+    if self_derived_bases.issubset(other_derived_bases):
+      # Avoid cycles on derived_from_tag. https://github.com/rwth-i6/returnn/issues/1054
+      return other.declare_same_as(self)
```

Comment:
I checked with this test case: 
```python
from returnn.tf.network import TFNetwork
from returnn.config import Config

n_in, n_out = 2, 3
config = Config({"extern_data": {"data": {"dim": n_in}}})
network = TFNetwork(config=config)
net_dict = {
  "output": {
    "class": "rec",
    "name_scope": "",
    "from": "data:data",
    "optimize_move_layers_out": False,
    "unit": {
      "layer1": {"class": "linear", "n_out": n_out, "from": "data:source", "is_output_layer": True},
      "layer2": {"class": "linear", "n_out": n_out, "from": "data:source", "is_output_layer": True,
                 "name_scope": "layer1"},
      "output": {"class": "combine", "kind": "add", "from": ["layer1", "layer2"]}
    }
  }
}
network.construct_from_dict(net_dict)
recusion_bug.py (END)
```

here the recursion is introduced for the `self` having tag `Dim{F'layer1:feature-dense'(3)}` and `other` having the same tag `Dim{F'layer1:feature-dense'(3)}` but the two dimensions are not the same object. The problem why this recursion is introduced is because `self is other` evaluates to `False` but `self == other` evaluates to `True` and thus the subset property is given. 

Is it required to have:
```python
if self is other:
  return
```
or can this be changed to:
```python
if self == other:
  return
```

Alternatively we could to the subset check with `is` instead of `==` like:
```python
other_derived_bases = other.get_derived_bases_list()
self_derived_bases = self.get_derived_bases_list()
subset = True
for self_dim in self_derived_bases:
  exists_in_base = False
  for other_dim in other_derived_bases:
    if self_dim is other_dim:
      exists_in_base = True 
      break
  if not exists_in_base:
    subset = False
    break
if subset:
  return other.declare_same_as(self)
```
This solves the recursion error and still solves the circular dependency problem

Reply:
Oh, right I see. But why do you introduce an extra line with `self == other`? Instead we could just change 'self is other` to `self == other` in line 898. If `self is other` is True than also `self == other`.

Reply:
> But why do you introduce an extra line with `self == other`? Instead we could just change `self is other` to `self == other` in line 898.

I did not see that here on GitHub. Right, this makes sense.